autowarefoundation/autoware.universe
# feat(elevation_map_loader): add support for seleceted_map_loader

Signed-off-by: Shin-kyoto <58775300+Shin-kyoto@users.noreply.github.com>

## Description



These PRs must be merged before this PR.

- https://github.com/autowarefoundation/autoware.universe/pull/3286
- https://github.com/autowarefoundation/autoware_launch/pull/285
- https://github.com/autowarefoundation/autoware_msgs/pull/57
   - add new service and msg

### Why?

- When you use large pointcloud map, map_loader cannot publish large map as one topic because of the limit size determined by [the default value in cyclonedds](https://github.com/eclipse-cyclonedds/cyclonedds/blob/09a02f5b2b9ba856d7d4c616510252cf9e508ec2/docs/manual/config/config_file_reference.rst#cycloneddsdomaininternalmaxsamplesize).
- So I want to add sequential map loading mode to `elevation_map_loader`. It is helpful to avoid exeeding the limit by using `selected_map_loader`.

### What?

- I would like to add `selected_map_loader` support to `elevation_map_loader`.
- In the default setting, `selected_map_loader` support is disabled. So if you want to use [this](https://github.com/Shin-kyoto/autoware.universe/commit/69b1e7b9cc2139858bef8922974b43b698a0e12a), please change your launcher like this.
- I assume that the each PCD file is not so large and the size of each PCD file does not exceed the limit size to receive(This is determined by [the default value in cyclonedds](https://github.com/eclipse-cyclonedds/cyclonedds/blob/09a02f5b2b9ba856d7d4c616510252cf9e508ec2/docs/manual/config/config_file_reference.rst#cycloneddsdomaininternalmaxsamplesize))

## Related links



- [TIER IV INTERNAL LINK](https://tier4.atlassian.net/browse/RT1-254)
- https://github.com/autowarefoundation/autoware.universe/pull/3286
- https://github.com/autowarefoundation/autoware_launch/pull/285
- https://github.com/autowarefoundation/autoware_msgs/pull/57


## Tests performed



I tested with a data from Autoware tutorial. The map is divided into 20m grids ([sample-map-rosbag_split.zip](https://github.com/autowarefoundation/autoware.universe/files/10349104/sample-map-rosbag_split.zip)).

I have confirmed that the Autoware performs the same as the current Autoware when use_differential_load is set false.

## Notes for reviewers



How to check

1. Check if these PRs are used in your environment.
   - https://github.com/autowarefoundation/autoware.universe/pull/3286
   - https://github.com/autowarefoundation/autoware_msgs/pull/56
2. Build autoware
3. Change your launcher like [this](Can you please output to console with throttle at about 5 second intervals?).
4. Run command below
```bash
ros2 launch autoware_launch logging_simulator.launch.xml map_path:=$HOME/data/map/sample-map-rosbag_split pointcloud_map_file:=pointcloud_map vehicle_model:=sample_vehicle sensor_model:=sample_sensor_kit planning:=false control:=false system:=false rviz:=true sensing:=false|grep -e elevation -e map
```
5. You should get the outputs like;
```bash
❯ ros2 launch autoware_launch logging_simulator.launch.xml map_path:=$HOME/data/map/sample-map-rosbag_split pointcloud_map_file:=pointcloud_map vehicle_model:=sample_vehicle sensor_model:=sample_sensor_kit planning:=false control:=false system:=false rviz:=true sensing:=false|grep -e elevation -e map
........
[component_container-3] [INFO 1679896837.303861415] [map.pointcloud_map_loader]: Load /home/shintarotomie/data/map/sample-map-rosbag_split/pointcloud_map/89400_42300.pcd (1 out of 23)
........
[component_container_mt-1] [INFO 1679896838.796591044] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: Create elevation map from pointcloud map 
[component_container_mt-1] [INFO 1679896839.235498083] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: Finish creating elevation map. Total time: 0.433 sec
[component_container_mt-1] [INFO 1679896839.666728529] [rosbag2_storage]: Opened database '/home/shintarotomie/work/2023/0317/autoware/install/elevation_map_loader/share/elevation_map_loader/data/elevation_maps/52fe0e6043a473758c9aab905232c887a7374632dabe1099c27b22effcbd93e1/52fe0e6043a473758c9aab905232c887a7374632dabe1099c27b22effcbd93e1_0.db3' for READ_WRITE.
[component_container_mt-1] [INFO 1679896839.672484128] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: Saving elevation map successful: true
```
6. Check rviz and you should see the elevation_map like;

![image](https://user-images.githubusercontent.com/58775300/227856478-811dfc18-b9d1-4e95-b57c-d4347599b161.png)

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.


## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.


## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.


After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/
